ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Block.hs

```diff
@@ -0,0 +1,242 @@
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+
+module Ouroboros.Consensus.Ledger.Byron.Block (
+    -- * Hash
+    ByronHash(..)
+  , mkByronHash
+    -- * Block
+  , ByronBlock(..)
+  , mkByronBlock
+  , annotateByronBlock
+    -- * Header
+  , Header(..)
+  , mkByronHeader
+  , byronBlockMatchesHeader
+    -- * Serialisation
+  , encodeByronBlock
+  , decodeByronBlock
+  , encodeByronHeader
+  , decodeByronHeader
+  , encodeByronHeaderHash
+  , decodeByronHeaderHash
+  ) where
+
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString.Lazy as Lazy
+import           Data.FingerTree.Strict (Measured (..))
+import           Data.Proxy
+import           Data.Typeable
+import           GHC.Generics (Generic)
+
+import           Codec.CBOR.Decoding (Decoder)
+import           Codec.CBOR.Encoding (Encoding)
+import qualified Codec.CBOR.Encoding as CBOR
+
+import           Cardano.Binary
+import           Cardano.Prelude (NoUnexpectedThunks (..))
+
+import qualified Cardano.Chain.Block as CC
+import qualified Cardano.Chain.Slotting as CC
+
+import           Ouroboros.Network.Block
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Ledger.Byron.Aux
+import           Ouroboros.Consensus.Ledger.Byron.Conversions
+import           Ouroboros.Consensus.Ledger.Byron.Orphans ()
+import           Ouroboros.Consensus.Util.Condense
+
+{-------------------------------------------------------------------------------
+  Header hash
+-------------------------------------------------------------------------------}
+
+newtype ByronHash = ByronHash { unByronHash :: CC.HeaderHash }
+  deriving stock   (Eq, Ord, Show, Generic)
+  deriving newtype (ToCBOR, FromCBOR, Condense)
+  deriving anyclass NoUnexpectedThunks
+
+mkByronHash :: ABlockOrBoundaryHdr ByteString -> ByronHash
+mkByronHash = \case
+    ABOBBoundaryHdr ebb -> ByronHash $ CC.boundaryHeaderHashAnnotated ebb
+    ABOBBlockHdr    mb  -> ByronHash $ CC.headerHashAnnotated mb
+
+{-------------------------------------------------------------------------------
+  Block
+-------------------------------------------------------------------------------}
+
+data ByronBlock = ByronBlock {
```

Comment:
Discuss why we cache the slot number (for EBBs) and the hash (because we need it all the time and it's expensive to compute).